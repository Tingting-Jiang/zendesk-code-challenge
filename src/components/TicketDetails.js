import React from "react";


// const TicketDetails = ({ticket = {
//     requester_id: 100,
//     subject: "in nostrud occaecat consectetur aliquip",
//     description: "Esse esse quis ut esse nisi tempor sunt. Proident",
//     priority: "low",
//     status: "open"
// }, setList}) => {


const TicketDetails = ({ticket, setList}) => {
    
    const {subject,
        description,
        priority,
        status,
        created_at,
        submitter_id,
        requester_id,
        assignee_id} = ticket;
   
    return (
        ticket && (
            <>
                <button
                    className="btn btn-primary my-5"
                    onClick={() => setList(true)}>
                    Back to Tickets  List
                </button>
                <h3>Subject: {subject === null ? "Unknown" : subject}</h3>
    
                <table className="table">
                    <thead>
                    <tr>
                        <th scope="col-2">#</th>
                        <th scope="col-3">Item</th>
                        <th scope="col-7">Details</th>
                       
                    </tr>
                    </thead>
                    
                    <tbody>


                    <tr>
                        <th scope="row">1</th>
                        <td>Description</td>
                        <td>{description === null ? "Unknown" : description}</td>
                    </tr>

                    <tr>
                        <th scope="row">2</th>
                        <td>Priority</td>
                        <td>{priority === null ? "Unknown" : priority}</td>
                    </tr>

                    <tr>
                        <th scope="row">3</th>
                        <td>Status</td>
                        <td>{status === null ? "Unknown" : status}</td>
                    </tr>
                    
                    <tr>
                        <th scope="row">4</th>
                        <td>Created At</td>
                        <td>{created_at === null ? "Unknown" : created_at}</td>
                    </tr>
                    <tr>
                        <th scope="row">5</th>
                        <td>Submitter ID</td>
                        <td>{submitter_id === null ? "Unknown" : submitter_id}</td>
                    </tr>
                    <tr>
                        <th scope="row">6</th>
                        <td>Requester ID</td>
                        <td>{requester_id === null ? "Unknown" : requester_id}</td>
                    </tr>

                    <tr>
                        <th scope="row">7</th>
                        <td>Assignee ID</td>
                        <td>{assignee_id === null ? "Unknown" : assignee_id}</td>
                    </tr>
                    
                    
                    </tbody>
            
                </table>
                
            
                {/*// <p>Status: {ticket.status}</p>*/}
                {/*// <p >Description: {ticket.description}</p>*/}
                {/*// <p>Priority: {ticket.priority} </p>*/}
            </>
        )
        
    );


};
export default TicketDetails;