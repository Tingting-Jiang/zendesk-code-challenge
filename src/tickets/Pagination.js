import React, { useState } from 'react'


const Pagination = ({ticketsPerPage, totalTickets, paginate}) =>{
    const pageNumbers = [];
    for (let i = 1; i <= Math.ceil(totalTickets / ticketsPerPage ); i++) {
        pageNumbers.push(i);
    };
    
    

    
    return (
        <nav className="nav-pills">
            <ul className="pagination">
                {pageNumbers.map(number => (
                    <li key={number} className="page-item"   >
                        <a onClick={() => {
                            paginate(number);
                            
                        }}
                            href="!#" className="page-link "
                      >
                            {number}
                        </a>
                    </li>
                ))}
            </ul>
        </nav>
    )
};
export default Pagination;


