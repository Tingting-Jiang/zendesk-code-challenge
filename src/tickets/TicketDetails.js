import React from "react";


const TicketDetails = ({ticket = {
    
    requester_id: 100,
    subject: "in nostrud occaecat consectetur aliquip",
    description: "Esse esse quis ut esse nisi tempor sunt. Proident",
    priority: "low",
    status: "open"
}, setList}) => {
    return (
        <>
            <button
                className="btn btn-primary my-5"
                onClick={() => setList(true)}>
                Back to Main Page
            </button>
            <h3>Subject: {ticket.subject}</h3>
            <p>Status: {ticket.status}</p>
            <p >Description: {ticket.description}</p>
            <p>Priority: {ticket.priority} </p>
        </>
    )


};
export default TicketDetails;