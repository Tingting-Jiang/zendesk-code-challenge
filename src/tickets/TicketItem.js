import React from 'react';



const TicketItem = ({ ticket}) => {
    
   
    return (
        <li className="list-group-item text-start">
            {/*<Link to=`/${ticket.requester_id}`>*/}
           
            {ticket.subject}
     
        </li>
    );
};
export default TicketItem;
