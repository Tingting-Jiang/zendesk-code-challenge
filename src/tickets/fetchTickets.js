// import URL from "../data/data";
// import TicketsPerPage from "../data/data";


const URL = "http://localhost:4000";
// const TicketsPerPage = 2;

export const fetchSomeTickets = (TicketsPerPage, page) =>
    
    fetch(`${URL}/${TicketsPerPage}/${page}`)
        .then(console.log("backend localhost: ", `${URL}/${TicketsPerPage}/${page}`))
        .then(response => response.json());

       
//
// export const fetchTickets = () =>
//     fetch(URL)
//         .then(response => response.json());



// export default {fetchTickets};
    
