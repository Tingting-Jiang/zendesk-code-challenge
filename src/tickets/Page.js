import React from "react";
import TicketItem from './TicketItem'
import tickets from '../service/data.json'

const TicketsPerPage = 25;


const Page = ({data}) =>{
    
    const dataNum = data.count();
    
    const NewPage = (count) => {
        const TotalPage = Math.floor(dataNum / TicketsPerPage) + (dataNum % TicketsPerPage > 0 ? 1 : 0);
        const total = [];
        for (let i = 1; i <= TotalPage; i++){
            total.push(i);
        }
        return(
            <>
                <ol className="list-group list-group-numbered">
                    {total.map((ticket, key) => {
                        return <TicketItem ticket={ticket} key={key} />;
                    })}
                </ol>
            </>
        )
        
    }
    
};
export default Page;


