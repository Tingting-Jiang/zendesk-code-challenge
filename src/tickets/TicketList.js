import React, { useEffect } from 'react'
import { fetchTickets } from './fetchTickets'

const TicketList = ({tickets, setList, setTicket}) => {
    
    // if (loading) {
    //     return <h2>Loading...</h2>;
    // }
    
    // console.log(tickets);
    
    
    const ViewTicket = (ticket) => {
        // console.log("event is", ticket);
        setTicket(ticket);
        setList(false);
        console.log("after setting: ", ticket);
        
    }

    return (
        
        <ul className="list-group mb-4">
            {tickets.map(ticketItem => (
                <li key={ticketItem.id}
                    className="list-group-item"
                onClick={() => ViewTicket(ticketItem)}>
                    {ticketItem.subject}
                    <span className="d-none">
                        {ticketItem.requester_id}
                    </span>
                
                </li>
            ))}
        </ul>
    
    )
}
export default TicketList;