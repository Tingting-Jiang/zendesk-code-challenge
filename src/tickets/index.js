import TicketItem from "./TicketItem";
// import tickets from "../service/data.json"
import { useEffect, useState } from 'react'
import fetchTickets from './fetchTickets'


const TicketList = () => {
  
    // const length = tickets.length;
    const [tickets, setTickets] = useState([]);
    useEffect(() =>
            fetchTickets()
                .then(test => {
                    setTickets(test.tickets);
                })
    ,[]);

    console.log("new data______", tickets);
    const length = tickets.length;
    

    return (
        <>
            <h3>{length} total tickets, 25 on this page</h3>
            <ol className="list-group list-group-numbered">
                {tickets.map((ticket, key) => {
                    return <TicketItem ticket={ticket} key={key} />;
                })}
            </ol>
        </>
    );
};
export default TicketList;
