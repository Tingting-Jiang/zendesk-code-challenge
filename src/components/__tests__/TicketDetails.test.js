import React from 'react';
import TicketDetails from '../TicketDetails'
import {shallow } from"enzyme";
import testData from "../../data/test.json";
import { cleanup } from '@testing-library/react';
import tickets from  "../../data/tickets.json";

describe("<TicketDetails/>", () =>{
    
    afterAll( () => {
        cleanup();
    })
    it("render TicketDetails without data", () =>{
        const wrapper = shallow(<TicketDetails/>);
        expect(wrapper).toMatchSnapshot();
    });
    
    it("render TicketDetails with fake data", () =>{
        const ticket = undefined;
        const setList = jest.fn();
        const wrapper = shallow(<TicketDetails ticket={ticket} setList={setList}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('#error-header').text()).toBe("Oops, the ticket is empty...");
        wrapper.find('#button2').simulate('click');
        expect(setList).toHaveBeenCalledTimes(1);
    });
    
    it("render TicketDetails with button click", () =>{
        const ticket = testData[0];
        const setList = jest.fn();
        const wrapper = shallow(<TicketDetails setList={setList}
                                               ticket={ticket}/>);
        wrapper.find('#button1').simulate('click');
        expect(setList).toHaveBeenCalledTimes(1);
    });
    
    it("render TicketDetails 1 data", () =>{
        const ticket = testData[0];
        const setList = jest.fn();
        const wrapper = shallow(<TicketDetails setList={setList}
                                                ticket={ticket}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("#detail-list").children()).toHaveLength(7);
        expect(wrapper.find('#header').text()).toBe("Subject: My printer is on fire!");
        expect(wrapper.find("#description").text()).toBe("Description: The smoke is very colorful.");

    })

    it("render TicketDetails 1 data", () =>{
        const ticket = tickets.tickets[24];
        const setList = jest.fn();
        const wrapper = shallow(<TicketDetails setList={setList}
                                               ticket={ticket}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('#header').text())
            .toBe("Subject: Lorem aliqua id nisi minim");
        expect(wrapper.find("#time").text())
            .toBe("Created At: 11/23/2021, 3:07:49 PM");

    })
    
    it("render TicketDetails 1 data", () =>{
        const ticket = testData[1];
        const setList = jest.fn();
        const wrapper = shallow(<TicketDetails setList={setList}
                                             ticket={ticket}/>);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('#header').text())
            .toBe("Subject: Unknown");
        expect(wrapper.find("#time").text())
            .toBe("Created At: 12/31/1969, 4:00:00 PM");
        
    })
})