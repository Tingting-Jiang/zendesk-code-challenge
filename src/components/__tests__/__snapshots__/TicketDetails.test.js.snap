// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketDetails/> render TicketDetails 1 data 1`] = `ShallowWrapper {}`;

exports[`<TicketDetails/> render TicketDetails 1 data 2`] = `ShallowWrapper {}`;

exports[`<TicketDetails/> render TicketDetails without data 1`] = `ShallowWrapper {}`;
